File: data-structure/trie/trie.go
package main

import "fmt"

// AlphatbetLetter index and letter
var AlphatbetLetter = map[int]string{
	0: "a", 1: "b", 2: "c", 3: "d", 4: "e", 5: "f", 6: "g", 7: "h",
	8: "i", 9: "j", 10: "k", 11: "l", 12: "m", 13: "n", 14: "o", 15: "p",
	16: "q", 17: "r", 18: "s", 19: "t", 20: "u", 21: "v", 22: "w",
	23: "x", 24: "y", 25: "z",
}

// Trie data structure shape
type Trie struct {
	children  []*Trie
	isEndWord bool
}

// Insert word to trie
func (t *Trie) Insert(word string) {
	root := t
	for i := 0; i < len(word); i++ {
		index := word[i] - 'a'
		if root.children[index] == nil {
			children := make([]*Trie, 26)
			root.children[index] = &Trie{children, false}
		}
		root = root.children[index]
	}
	root.isEndWord = true
}

// Search word
func (t *Trie) Search(word string) bool {
	root := t
	// loop char in string
	// each char convert char to index
	// check existing in trie and end of the world
	for i := 0; i < len(word); i++ {
		index := word[i] - 'a'
		if root.children[index] == nil {
			return false
		}
		root = root.children[index]
	}
	return root != nil && root.isEndWord
}

// IsEmpty check node does not have children
func (t *Trie) IsEmpty() bool {
	root := t
	for i := 0; i < 26; i++ {
		if root.children[i] != nil {
			return false
		}
	}
	return true
}

// Remove word
func (t *Trie) Remove(word string, depth int) {
	if t == nil {
		return
	}

	if len(word) == depth {
		if t.isEndWord == true {
			t.isEndWord = false
		}

		/*
		 * If this is the last node
		 */
		if t.IsEmpty() == true {
			t = nil
			return
		}
	}

	/*
	 * If the current node has no child and it's not the final letter of the world
	 * Why?
	 * - Because the previous node was delete => now the parent does not have any child
	 */
	if t.isEndWord == false && t.IsEmpty() == true {
		t = nil
		return
	}

	/*
	 * Depth here is the level of tree
	 * The length of the branch is the length of the word
	 */
	index := word[depth] - 'a'

	/*
	 * Recursive call to access levels of the tree
	 * Each time it calls, it will move down 1 level
	 * Based on the "depth" we can access the letter of the word
	 */
	t.children[index].Remove(word, depth+1)
}

// GetSuggestionItem recursive call to get the suggestion item
func (t *Trie) GetSuggestionItem(keyword string, result *[]string) {
	if t.isEndWord == true {
		*result = append((*result), keyword)
	}

	/*
	 * With every node, we check 26 letters has value or not
	 * if it has value, recursive call GetSuggestionItem, with
	 * - keyword update
	 * - the list of suggestion items
	 */
	for i := 0; i < 26; i++ {
		if t.children[i] != nil {
			letter := AlphatbetLetter[i]
			t.children[i].GetSuggestionItem(keyword+letter, result)
		}
	}
}

// GetSuggestion return list of suggestion items
func (t *Trie) GetSuggestion(keyword string) []string {
	/*
	 * Go to the Node contains the last key of keyword
	 * if the keyword is not existing => return []
	 * otherwise call GetSuggestionItem
	 */

	root := t
	for i := 0; i < len(keyword); i++ {
		index := keyword[i] - 'a'
		if root.children[index] == nil {
			return nil
		}
		root = root.children[index]
	}
	var result []string
	root.GetSuggestionItem(keyword, &result)
	return result
}

func main() {
	children := make([]*Trie, 26)
	trie := Trie{children, false}
	trie.Insert("the")
	trie.Insert("there")
	trie.Insert("these")
	trie.Insert("their")

	// existing := trie.Search("their")
	// fmt.Println(existing)

	// trie.Remove("their", 0)

	// existing1 := trie.Search("their")
	// fmt.Println(existing1)

	suggestion := trie.GetSuggestion("the")
	fmt.Println(suggestion)
}

File: data-structure/trie/trie.py
alphatbet = 26

alphatbet_mapping = {
  0: "a", 1: "b", 2: "c", 3: "d", 4: "e", 5: "f", 6: "g", 7: "h",
  8: "i", 9: "j", 10: "k", 11: "l", 12: "m", 13: "n", 14: "o", 15: "p",
  16: "q", 17: "r", 18: "s", 19: "t", 20: "u", 21: "v", 22: "w",
  23: "x", 24: "y", 25: "z"
}

class TrieNode:
  def __init__(self):
    self.children = [None] * alphatbet
    self.is_end_of_word = False

  def insert(self, keyword):
    for c in keyword:
      index = ord(c) - ord("a");

      if not self.children[index]:
        self.children[index] = TrieNode()
      self = self.children[index]

    self.is_end_of_word = True

  def is_last_node(self):
    for i in range(alphatbet):
      if self.children[i]:
        return False
    return True

  def search(self, keyword):
    for c in keyword:
      index = ord(c) - ord("a");

      if not self.children[index]:
        return False

      self = self.children[index]
    return self and self.is_end_of_word

  def get_auto_complete(self, prefix, result):
    if self.is_end_of_word:
      result.append(prefix)

    for i in range(alphatbet):
      if self.children[i]:
        self.children[i].get_auto_complete(prefix + alphatbet_mapping[i], result)

  def get_search_suggestion(self, prefix):
    root = self
    for c in prefix:
      index = ord(c) - ord("a")
      if not root.children[index]:
        return []
      root = root.children[index]

    list_keywords = []
    root.get_auto_complete(prefix, list_keywords)

    return list_keywords

  def is_empty(self):
    for i in range(alphatbet):
      if self.children[i]:
        return False
    return True

  def remove_word(self, word, depth = 0):
    # if current node has children => set node to false
    # if current node has no child => delete it
    ## recursive to the node above, if parent node has no child => delete it
    if not self:
      return None
    
    if len(word) == depth:
      if self.is_end_of_word == True:
        self.is_end_of_word = False
      
      if self.is_empty():
        self = None
      return True

    if self.is_empty() and self.is_end_of_word == False:
      self = None
      return True

    index = ord(word[depth]) - ord("a");
    self.children[index].remove_word(word, depth + 1)


trie = TrieNode()
trie.insert("the")
trie.insert("there")
trie.insert("these")
trie.insert("their")

# trie.insert("a")
# trie.insert("answer")
# trie.insert("any")

# trie.insert("by")
# trie.insert("bye")

# trie.insert("ab")

# result = trie.search("ab")
# print result

# trie = trie.remove_word("ab")

# result1 = trie.search("ab")
# print trie.children[0]

# result1 = trie.search("there")
# print result1

# result = trie.search("the")
# print result
# result1 = trie.search("theire")
# print result1

# suggestion = trie.get_search_suggestion("a")
# print suggestion

ab = trie.remove_word("there")
print "ab ab", ab

suggestion = trie.get_search_suggestion("the")
print suggestion